doc/man7/provider-keyexch.pod

```diff
@@ -0,0 +1,180 @@
+
+ typedef void *(OSSL_OP_keyexch_newctx_fn)(void *provctx);
+ static ossl_inline OSSL_OP_keyexch_newctx_fn
+     OSSL_get_OP_keyexch_newctx(const OSSL_DISPATCH *opf);
+
+B<OSSL_DISPATCH> arrays are indexed by numbers that are provided as
+macros in L<openssl-core_numbers.h(7)>, as follows:
+
+ OP_keyexch_newctx           OSSL_FUNC_KEYEXCH_NEWCTX
+ OP_keyexch_freectx          OSSL_FUNC_KEYEXCH_FREECTX
+ OP_keyexch_dupctx           OSSL_FUNC_KEYEXCH_DUPCTX
+
+ OP_keyexch_init             OSSL_FUNC_KEYEXCH_INIT
+ OP_keyexch_set_peer         OSSL_FUNC_KEYEXCH_SET_PEER
+ OP_keyexch_derive           OSSL_FUNC_KEYEXCH_DERIVE
+
+ OP_keyexch_set_params       OSSL_FUNC_KEYEXCH_SET_PARAMS
+
+A key exchange algorithm implementation may not implement all of these functions.
+In order to be a consistent set of functions a provider must implement
+OP_keyexch_newctx, OP_keyexch_freectx, OP_keyexch_init and OP_keyexch_derive.
+All other functions are optional.
+
+A key exchange algorithm must also implement some mechanism for generating,
+loading or importing keys via the key management (OSSL_OP_KEYMGMT) operation.
+See L<provider-keymgmt(7)> for further details.
+
+=head2 Context Management Functions
+
+OP_keyexch_newctx() should create and return a pointer to a provider side
+structure for holding context information during a key exchange operation.
+A pointer to this context will be passed back in a number of the other key
+exchange operation function calls.
+The paramater B<provctx> is the provider context generated during provider
+initialisation (see L<provider(3)>).
+
+OP_keyexch_freectx() is passed a pointer to the provider side key exchange
+context in the B<ctx> parameter.
+This function should free any resources associated with that context.
+
+OP_keyexch_dupctx() should duplicate the provider side key exchange context in
+the B<ctx> parameter and return the duplicate copy.
+
+=head2 Shared Secret Derivation Functions
+
+OP_keyexch_init() initialises a key exchange operation given a newly created
+provider side key exchange context in the B<ctx> paramter, and a pointer to a
+provider key object in the B<provkey> parameter.
+The key object will have been previously generated, loaded or imported into the
+provider using the key management (OSSL_OP_KEYMGMT) operation (see
+provider-keymgmt(7)>.
+
+OP_keyexch_set_peer() is called to supply the peer's public key (in the
+B<provkey> parameter) to be used when deriving the shared secret.
+It is also passed a previously initialised key exchange context in the B<ctx>
+parameter.
+The key object will have been previously generated, loaded or imported into the
+provider using the key management (OSSL_OP_KEYMGMT) operation (see
+provider-keymgmt(7)>.
+
+OP_keyexch_derive() performs the actual key exchange itself by deriving a shared
+secret.
+A previously initialised key exchange context is passed in the B<ctx>
+parameter.
+The derived secret should be written to the location B<secret> which should not
```

Comment:
s/should/will/ ??
s/which should/but will/

Reply:
This is talking about the provider responsibilities when implementing these functions. "Should" seems to make more sense to me in that context.